api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -370,6 +382,49 @@ private void streamToOutput(final boolean streamTimeoutReached) {
         }
     }
 
+    private void sendToDeadLetterQueue(final ConsumedEvent event, final int failedCommitsCount) {
+        // TODO: how do we handle AVRO here?
+
+        final String failedEventString = new String(event.getEvent(), StandardCharsets.UTF_8);
+        final JSONObject failedEvent = new JSONObject(failedEventString);
+
+        final JSONObject errorMessage =
+                new JSONObject()
+                .put("message", "skipped due to failed commits count: " + failedCommitsCount);
+
+        final SubscriptionCursorWithoutToken cursor =
+                getContext().getCursorConverter().convertToNoToken(event.getPosition());
+
+        final JSONObject deadLetter =
+                new JSONObject()
+                .put("subscription_id", getContext().getSubscription().getId())
+                .put("event_type", cursor.getEventType())
+                .put("partition", cursor.getPartition())
+                .put("offset", cursor.getOffset())
+                .put("error", errorMessage)
+                .put("event", failedEvent)
+                .put("metadata",
+                        new JSONObject()
+                        .put("eid", getContext().getUuidGenerator().randomUUID().toString())
+                        .put("occurred_at", Instant.now().toString()));
+
+        final JSONArray deadLetterBatch =
+                new JSONArray().put(deadLetter);
+
+        final EventPublishResult result;
+        try {
+            result = getContext().getEventPublisher().publishInternal(
+                    deadLetterBatch.toString(), getContext().getDeadLetterQueueEventTypeName(), null);
+        } catch (final Exception e) {
+            throw new DeadLetterQueueStoreException("Failed to store an event to the dead letter queue", e);
+        }
+
+        if (result.getStatus() != EventPublishingStatus.SUBMITTED) {
+            throw new DeadLetterQueueStoreException("Failed to store an event to the dead letter queue: " +
```

Comment:
it looks like it blocks consumption of the event type in case there an issue with completely unrelated event type. we need to rethink it and move into direction of non blocking processing.

Reply:
> it looks like it blocks consumption of the event type in case there an issue with completely unrelated event type

Yes, but only if Auto-DLQ is used and there actually some events that need to be stored, the likelihood of that coinciding with any issues with the DLQ store event type is pretty slim.

> we need to rethink it and move into direction of non blocking processing.

I don't think that's possible at all.  The whole point is to avoid data loss.